File: airbyte-commons-worker/src/test/java/io/airbyte/workers/internal/DestinationTimeoutMonitorTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal;

import static io.airbyte.metrics.lib.OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT;
import static io.airbyte.metrics.lib.OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import io.airbyte.metrics.lib.MetricAttribute;
import io.airbyte.metrics.lib.MetricClient;
import java.time.Duration;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import org.junit.jupiter.api.Test;

class DestinationTimeoutMonitorTest {

  private final MetricClient metricClient = mock(MetricClient.class);

  @Test
  void testNoTimeout() {
    DestinationTimeoutMonitor destinationTimeoutMonitor = new DestinationTimeoutMonitor(
        UUID.randomUUID(),
        UUID.randomUUID(),
        metricClient,
        Duration.ofMinutes(5),
        true,
        Duration.ofMillis(500));

    destinationTimeoutMonitor.startAcceptTimer();
    destinationTimeoutMonitor.startNotifyEndOfInputTimer();

    assertDoesNotThrow(() -> destinationTimeoutMonitor.runWithTimeoutThread(CompletableFuture.runAsync(() -> {
      try {
        Thread.sleep(2000);
      } catch (final InterruptedException e) {
        throw new RuntimeException(e);
      }
    })));

    verify(metricClient, never()).count(eq(WORKER_DESTINATION_ACCEPT_TIMEOUT), anyLong(), any(MetricAttribute.class));
    verify(metricClient, never()).count(eq(WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT), anyLong(), any(MetricAttribute.class));
  }

  @Test
  void testAcceptTimeout() {
    DestinationTimeoutMonitor destinationTimeoutMonitor = new DestinationTimeoutMonitor(
        UUID.randomUUID(),
        UUID.randomUUID(),
        metricClient,
        Duration.ofSeconds(1),
        true,
        Duration.ofSeconds(1));

    destinationTimeoutMonitor.startAcceptTimer();

    assertThrows(
        DestinationTimeoutMonitor.TimeoutException.class,
        () -> destinationTimeoutMonitor.runWithTimeoutThread(CompletableFuture.runAsync(() -> {
          try {
            Thread.sleep(Long.MAX_VALUE);
          } catch (final InterruptedException e) {
            throw new RuntimeException(e);
          }
        })));

    verify(metricClient).count(eq(WORKER_DESTINATION_ACCEPT_TIMEOUT), eq(1L), any(MetricAttribute.class));
    verify(metricClient, never()).count(eq(WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT), anyLong(), any(MetricAttribute.class));
  }

  @Test
  void testNotifyEndOfInputTimeout() {
    DestinationTimeoutMonitor destinationTimeoutMonitor = new DestinationTimeoutMonitor(
        UUID.randomUUID(),
        UUID.randomUUID(),
        metricClient,
        Duration.ofSeconds(1),
        true,
        Duration.ofSeconds(1));

    destinationTimeoutMonitor.startNotifyEndOfInputTimer();

    assertThrows(
        DestinationTimeoutMonitor.TimeoutException.class,
        () -> destinationTimeoutMonitor.runWithTimeoutThread(CompletableFuture.runAsync(() -> {
          try {
            Thread.sleep(Long.MAX_VALUE);
          } catch (final InterruptedException e) {
            throw new RuntimeException(e);
          }
        })));

    verify(metricClient, never()).count(eq(WORKER_DESTINATION_ACCEPT_TIMEOUT), anyLong(), any(MetricAttribute.class));
    verify(metricClient).count(eq(WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT), eq(1L), any(MetricAttribute.class));
  }

  @Test
  void testTimeoutNoException() {
    DestinationTimeoutMonitor destinationTimeoutMonitor = new DestinationTimeoutMonitor(
        UUID.randomUUID(),
        UUID.randomUUID(),
        metricClient,
        Duration.ofSeconds(1),
        false,
        Duration.ofSeconds(1));

    destinationTimeoutMonitor.startAcceptTimer();

    assertDoesNotThrow(
        () -> destinationTimeoutMonitor.runWithTimeoutThread(CompletableFuture.runAsync(() -> {
          try {
            Thread.sleep(Long.MAX_VALUE);
          } catch (final InterruptedException e) {
            throw new RuntimeException(e);
          }
        })));

    verify(metricClient).count(eq(WORKER_DESTINATION_ACCEPT_TIMEOUT), eq(1L), any(MetricAttribute.class));
    verify(metricClient, never()).count(eq(WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT), anyLong(), any(MetricAttribute.class));
  }

}

File: airbyte-container-orchestrator/src/test/kotlin/io/airbyte/container/orchestrator/worker/io/DestinationTimeoutMonitorTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker.io

import io.airbyte.container.orchestrator.worker.context.ReplicationInputFeatureFlagReader
import io.airbyte.featureflag.DestinationTimeoutSeconds
import io.airbyte.featureflag.ShouldFailSyncOnDestinationTimeout
import io.airbyte.metrics.MetricAttribute
import io.airbyte.metrics.MetricClient
import io.airbyte.metrics.OssMetricsRegistry
import io.airbyte.persistence.job.models.ReplicationInput
import io.micrometer.core.instrument.Counter
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.time.Duration
import java.util.UUID

internal class DestinationTimeoutMonitorTest {
  private val metricClient: MetricClient = mockk()

  @Test
  fun testNoTimeout() {
    every {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    } returns mockk<Counter>()
    every {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    } returns mockk<Counter>()

    val replicationInputFeatureFlagReader: ReplicationInputFeatureFlagReader =
      mockk {
        every { read(DestinationTimeoutSeconds) } returns Duration.ofMinutes(5).toSeconds().toInt()
        every { read(ShouldFailSyncOnDestinationTimeout) } returns true
      }
    val replicationInput = ReplicationInput().withWorkspaceId(UUID.randomUUID()).withConnectionId(UUID.randomUUID())

    val destinationTimeoutMonitor =
      DestinationTimeoutMonitor(
        replicationInput = replicationInput,
        replicationInputFeatureFlagReader = replicationInputFeatureFlagReader,
        metricClient = metricClient,
      )

    destinationTimeoutMonitor.startAcceptTimer()
    destinationTimeoutMonitor.startNotifyEndOfInputTimer()
    Thread.sleep(10)
    assertFalse(destinationTimeoutMonitor.hasTimedOut())

    verify(exactly = 0) {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    }
    verify(exactly = 0) {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    }
  }

  @Test
  fun testAcceptTimeout() {
    every {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    } returns mockk<Counter>()
    every {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    } returns mockk<Counter>()

    val replicationInputFeatureFlagReader: ReplicationInputFeatureFlagReader =
      mockk {
        every { read(DestinationTimeoutSeconds) } returns Duration.ofMillis(1).toSeconds().toInt()
        every { read(ShouldFailSyncOnDestinationTimeout) } returns true
      }

    val replicationInput = ReplicationInput().withWorkspaceId(UUID.randomUUID()).withConnectionId(UUID.randomUUID())

    val destinationTimeoutMonitor =
      DestinationTimeoutMonitor(
        replicationInput = replicationInput,
        replicationInputFeatureFlagReader = replicationInputFeatureFlagReader,
        metricClient = metricClient,
      )

    destinationTimeoutMonitor.startAcceptTimer()
    Thread.sleep(10)
    assertTrue(destinationTimeoutMonitor.hasTimedOut())

    verify(exactly = 1) {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    }
    verify(exactly = 0) {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    }
  }

  @Test
  fun testNotifyEndOfInputTimeout() {
    every {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    } returns mockk<Counter>()
    every {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    } returns mockk<Counter>()

    val replicationInputFeatureFlagReader: ReplicationInputFeatureFlagReader =
      mockk {
        every { read(DestinationTimeoutSeconds) } returns Duration.ofMillis(1).toSeconds().toInt()
        every { read(ShouldFailSyncOnDestinationTimeout) } returns true
      }
    val replicationInput = ReplicationInput().withWorkspaceId(UUID.randomUUID()).withConnectionId(UUID.randomUUID())

    val destinationTimeoutMonitor =
      DestinationTimeoutMonitor(
        replicationInput = replicationInput,
        replicationInputFeatureFlagReader = replicationInputFeatureFlagReader,
        metricClient = metricClient,
      )

    destinationTimeoutMonitor.startNotifyEndOfInputTimer()
    Thread.sleep(10)
    assertTrue(destinationTimeoutMonitor.hasTimedOut())

    verify(exactly = 0) {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    }
    verify(exactly = 1) {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    }
  }

  @Test
  fun testTimeoutNoException() {
    every {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    } returns mockk<Counter>()
    every {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    } returns mockk<Counter>()

    val replicationInputFeatureFlagReader: ReplicationInputFeatureFlagReader =
      mockk {
        every { read(DestinationTimeoutSeconds) } returns Duration.ofMillis(5).toSeconds().toInt()
        every { read(ShouldFailSyncOnDestinationTimeout) } returns false
      }
    val replicationInput = ReplicationInput().withWorkspaceId(UUID.randomUUID()).withConnectionId(UUID.randomUUID())

    val destinationTimeoutMonitor =
      DestinationTimeoutMonitor(
        replicationInput = replicationInput,
        replicationInputFeatureFlagReader = replicationInputFeatureFlagReader,
        metricClient = metricClient,
      )

    destinationTimeoutMonitor.startAcceptTimer()
    Thread.sleep(10)
    assertFalse(destinationTimeoutMonitor.hasTimedOut())

    verify(exactly = 0) {
      metricClient.count(metric = OssMetricsRegistry.WORKER_DESTINATION_ACCEPT_TIMEOUT, value = any(), attributes = anyVararg<MetricAttribute>())
    }
    verify(exactly = 0) {
      metricClient.count(
        metric = OssMetricsRegistry.WORKER_DESTINATION_NOTIFY_END_OF_INPUT_TIMEOUT,
        value = any(),
        attributes = anyVararg<MetricAttribute>(),
      )
    }
  }
}
